File: chatclient.py
#!/usr/bin/python

#	Name: Andrew Michael Calhoun
#	Class: CS372 Introduction to Networks
#	Project 1
#	Filename: chatclient.py

# Dependencies
from struct import *  # https://docs.python.org/2/library/struct.html
import os
import getopt
import signal
import sys
import socket


# FUNCTIONS

# takes a hostname and port number
# opens a socket and connects to host on the port
# socket file descriptor

def client_connect(host, port):
	if port < 1024 or port > 65535:
		print "Ports must be in range 1024-65535" # Ports less than 1024 may conflict, and greater than 65535 do not work because 16-bit overflow.
		sys.exit(0)

	# open and connect to a socket
	clientSocket = socket.socket(socket.AF_INET,socket.SOCK_DGRAM)
	
	try:
		clientSocket.client_connect((host, port))
	
	except: 
		print "Error connecting to " + host + " : " + str(port)
		sys.exit(0)

	return clientSocket

	# Based off Socket Programming Primer Lecture

	# sentence = raw_input('Input lowercase sentence:')
	# clientSocket.send(sentence)
	# modifiedSentence = clientSocket.recv(1024)

	# print 'From Server:', modifiedSentence
	# clientSocket.close()

# From https://docs.python.org/2/library/struct.html

def sendMessage(sockfd, outMsg):
	size = len(outgoing) 	# length of the message
	data = pack('I', size)	# http: https://docs.python.org/2/library/struct.html
							# 's' is string size
	sockfd.send(data)		# size of the message
	sockfd.send(outgoing)	# send the message itself


def receiveMsg(sockfd):
	data = sockfd.recv(4)
	recvsize = unpack('I', data)
	incoming = sockfd.recv(recvsize[0])
	return incoming

def getUserName():
	validUserName = False
	while validUserName == False:
		validUserName = True
		print "Enter your user name, one word, up to 16 characters.\nPlease avoid using spaces or special characters."
		handle = raw_input()
		for characters in handle:
			if character == ' ':
			# if character ==  ' ' || '!' || '@' || '#' || '$' || '%' || '^' || '&' || '*' || '(' || ')' || '-' || '=' || '_' || '+' || '[' || ']' || '|' || '\\' || '{' || '}' || ';' || ':' || '\'' || '"' || '<' || '>' || '/' || ',' || '.' || '/' || '?':
				validUserName = False
				print("Invalid User Name: May not contain spaces or special characters.")
			if len(handle) > 16:
				validUserNAme = False
				print("Invalid User Name: Must be 16 characters or less.")
		return userName



# MAIN

# cmd line args
if len(sys.argv) != 3:
	print "\nUsage:\n $ python chatclient.py [hostname] [port number]\n\n" 
else:
	# connect to the server -- no server... no worky.
	serverConnect = client_connect(sys.argv[1], int(sys.argv[2]))

	# get the user name
	userName = getUseName()

	quit = False
	while quit == False:
		print (userName + "$"),
		message = raw_input()
		if message == "\quit":
			quit = True

	if quit == True:
		outMsg = "\quit"
	else:
		message = (userName + " > " + message) # append to username.

		sendMessage(serverConnect, message)

	if quit == False:
		incMsg = receiveMsg(clientSocket);
		if incMsg == "quit":
			printf("Connection reset by server; exiting.\n")
			clientSocket.close()
			quit = True
		else:
			print incMsg

	else:
		clientSocket.close()


#if __name__ == '__main__':
#	client_connect(host, port)

File: chatclient.c

/* Andrew M. Calhoun
calhouna@oregonstate.edu
Due Date: 2/7/2016
Project I 
chatserver.cpp


// Some Patterns are based off of Beej's guide: 
// http://beej.us/guide/bgnet/output/html/multipage/clientserver.html#simpleserver

Multithreading based off of: 
// http://www.cplusplus.com/forum/unices/116977/


Chat Server Defines and Includes */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <fcntl.h>
#include <netdb.h> 

#define MSGBUFFER 500
#define HANDLE 64

// This function is called when a system call fails 
void errorMsg(const char *msg)
{
    printf("Something went wrong.\n");
    printf("%s\n", msg);
    exit(1);
}

int main(int argc, char *argv[])
{

    // Program variables:
    // Set up some variables for the program
    // hostName     : to store the hostname from command line
    // buffer: to store the user enter for a message
    // outgoingMsg : to store the final constructed message (userName + outgoingMsg) 
    // userName      : to store the user name and set it a prompt
    // quit  :stores /quit for quitting
    // MessageLength: to store the messages lenghts
    // status       : varibale to set flag to continue a session between client and server
    // i            : index for loop


	//Initialize the socket struct
    int sockfd, portNum;

  //  const char *host = "localhost";

    struct sockaddr_in server_address;
    struct hostent *server;
    const char *hostname = "localhost";

    char hostName[HANDLE];
    char outgoingMsg[MSGBUFFER];
    char buffer[MSGBUFFER];
    char userName[HANDLE];
    size_t messageLength;
    int i, n, status = 0;
   
    char *quit = (char*) malloc(4);

    if (argc < 3) {
       fprintf(stderr, "Please provide host or port.\nUsage: $ %s Hostname port\n", argv[0]);
       exit(1);
    }

    // Get the host name and the port from the command line
    sprintf(hostName,"%s",argv[1]);
    portNum = atoi(argv[2]);

    // Get the user name
	printf("Please enter your name (no spaces): ");
	fgets(userName, HANDLE, stdin);
	size_t length = strlen(userName) - 1;
	if (userName[length] == '\n')
			userName[length] = '\0';

	strcat(userName, " >>");

	// Creatig the socket  
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) 
        error("Error opening socket");

    // Get the address of the host
    server = gethostbyname(hostName);
    if (server == NULL) {
        fprintf(stderr,"Error: no such host\n");
        exit(1);
    }

    // Initlize the server address
    bzero((char *) &server_address, sizeof(server_address));
    server_address.sin_family = AF_INET;
    bcopy((char *)server->h_addr, (char *)&server_address.sin_addr.s_addr, server->h_length);
    server_address.sin_port = htons(portNum);

    // Connect to the server
    if (connect(sockfd,(struct sockaddr *) &server_address,sizeof(server_address)) < 0) 
        errorMsg("Error: connecting to server.");

    // Send the start message to the server
    n = write(sockfd,"PORTNUM",7);
    if (n < 0) 
         errorMsg("Error: writing to socket");

    // While there is a connection keep sending and receiving messages
    while(status == 0) {

    	// display userName and get the user message
    	printf("%s ", userName);
    	bzero(buffer,500);
	    fgets(buffer, 500, stdin);

	    messageLength = strlen(buffer) - 1;
		if (buffer[messageLength] == '\n')
    			buffer[messageLength] = '\0';

    	// Construct the message + userName to be sent to server
    	bzero(outgoingMsg,500);
    	strcat(outgoingMsg, userName);
    	strcat(outgoingMsg, " ");
    	strcat(outgoingMsg, buffer);

    	// send the message to the server
    	n = write(sockfd,outgoingMsg,500);
    	if (n < 0) 
	         errorMsg("ERROR writing to socket");

	    // Check if the user sent a "quit" message to exit connection
	    messageLength = strlen(outgoingMsg);
		for (i=0; i< messageLength ; i++)
		{
			if (outgoingMsg[i] == '>')
				break;
		}
		i = i + 3;
		bzero(quit,5);
		strncpy(quit, outgoingMsg + i , 5);
		if (strcmp(quit,"\\quit") == 0)
		{
			printf ("Client sent a quit message! will quit now!\n");
			status = 1;
			break;
		}

		// If the message is not "quit" then continue the connection
	    bzero(buffer,500);
	    bzero(quit,5);
	    n = read(sockfd,buffer,500);
	    if (n < 0)
			errorMsg("Error reading from socket");

		// Check if the server sent a "quit" message to end connection
		strncpy(quit, buffer + 10 , 5);

		if (strcmp(quit,"\\quit") == 0)
		{
			printf ("Server Sent a quit message! will quit now!\n");
			status = 1;
			break;
		}
		// if not display the message
    	printf ("%s\n",buffer);
    }
    
   	// freeing the memory from quit allocation
    free(quit);
	   
    // Close the socket / end of the request
    close(sockfd);
    
    return 0;
}